### PR TITLE
[API] Not being able to modify cart after completed checkout

### PR DESCRIPTION
**Pull request by arti0090**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: features/checkout/preventing_cart_to_be_modified_after_checkout.feature [hunk 1]
```diff
@@ -0,0 +1,71 @@
+@checkout
+Feature: Preventing cart to be modified after checkout by Customer
+    In order to not modify cart data after checkout
+    As a Customer
+    I want my cart to blocked from making changes after checkout
```
Comment (AdamKasp) [inline features/checkout/preventing_cart_to_be_modified_after_checkout.feature:5]:
```suggestion
Feature: Preventing cart to be modified after checkout
    In order to have order immutable after checkout
    As a Customer
    I want to not be able to modify placed order
```
WDYT?

File: features/checkout/preventing_cart_to_be_modified_after_checkout.feature [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a product "Sig Sauer P226" priced at "$499.99"
+        And the store has a product "AK-47" priced at "$99.99"
+        And the store ships everywhere for free
+        And the store ships everywhere with ups
+        And the store allows paying with "Cash on Delivery"
+        And the store allows paying with "Helicopter Money"
```
Comment (AdamKasp) [inline features/checkout/preventing_cart_to_be_modified_after_checkout.feature:14]:
```suggestion
        And the store allows paying also with "Helicopter Money"
```

File: features/checkout/preventing_cart_to_be_modified_after_checkout.feature [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        And I am a logged in customer
+
+    @api
+    Scenario: Changing address information after completing checkout
```
Comment (AdamKasp) [inline features/checkout/preventing_cart_to_be_modified_after_checkout.feature:18]:
```suggestion
    Scenario: Changing billing address after checkout
```

imo `completing` is redundant here and the following scenarios

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 4]
```diff
@@ -688,6 +705,43 @@ public function iShouldBeInformedThatThisPaymentMethodHasBeenDisabled(PaymentMet
...
+     */
+    public function iTryToRemoveProductFromTheCart(ProductInterface $product, string $tokenValue): void
+    {
+        $this->removeOrderItemFromCart($product->getId(), $tokenValue);
+    }
+
+    /**
+     * @When /^I try to change quantity to (\d+) of (product "[^"]+") from the (cart)$/
+     */
+    public function iTryToChangeQuantityToOfProductFromTheCart(int $quantity, ProductInterface $product, string $tokenValue)
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:727]:
```suggestion
    public function iTryToChangeQuantityToOfProductFromTheCart(int $quantity, ProductInterface $product, string $tokenValue): void
```

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 5]
```diff
@@ -688,6 +705,43 @@ public function iShouldBeInformedThatThisPaymentMethodHasBeenDisabled(PaymentMet
...
+    {
+        $this->putProductToCart($product, $tokenValue, $quantity);
+    }
+
+    /**
+     * @Then /^I should be informed that cart is no longer available$/
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:733]:
```suggestion
     * @Then I should be informed that cart is no longer available
```

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 6]
```diff
@@ -62,6 +65,12 @@ final class CheckoutContext implements Context
     /** @var RepositoryInterface */
     private $paymentMethodRepository;
 
+    /** @var AdminToShopIriConverterInterface */
+    private $adminToShopIriConverter;
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:69]: if I see correctly you don't use it anywhere.

File: src/Sylius/Behat/Context/Setup/ShippingContext.php [hunk 7]
```diff
@@ -93,15 +88,16 @@ public function storeShipsEverythingForFree(ZoneInterface $zone): void
     }
 
     /**
-     * @Given the store ships everywhere for free
+     * @Given the store ships everywhere for :shippingName
+     * @Given the store ships everywhere with :shippingName
      */
-    public function theStoreShipsEverywhereForFree(): void
+    public function theStoreShipsEverywhereWith(string $shippingName): void
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Setup/ShippingContext.php:94]:
```suggestion
    public function theStoreShipsEverywhereWith(string $shippingMethodName): void
```

**Review by AdamKasp — CHANGES_REQUESTED** [review]

File: features/checkout/preventing_cart_to_be_modified_after_checkout.feature [hunk 8]
```diff
@@ -0,0 +1,71 @@
+@checkout
+Feature: Preventing cart to be modified after checkout by Customer
```
Comment (GSadee) [inline features/checkout/preventing_cart_to_be_modified_after_checkout.feature:2]:
```suggestion
Feature: Preventing cart from being modified after checkout by Customer
```

File: features/checkout/preventing_cart_to_be_modified_after_checkout.feature (hunk 1 above)
Comment (GSadee) [reply features/checkout/preventing_cart_to_be_modified_after_checkout.feature:5]: But `I want not to be able to modify placed order`

File: features/checkout/preventing_cart_to_be_modified_after_checkout.feature (hunk 3 above)
Comment (GSadee) [inline features/checkout/preventing_cart_to_be_modified_after_checkout.feature:18]:
```suggestion
    Scenario: Preventing from changing billing address after checkout
```

File: features/checkout/preventing_cart_to_be_modified_after_checkout.feature [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        And the store allows paying with "Cash on Delivery"
+        And the store allows paying with "Helicopter Money"
+        And I am a logged in customer
+
+    @api
+    Scenario: Changing address information after completing checkout
+        Given I have product "Sig Sauer P226" in the cart
+        And I specified the billing address as "Ankh Morpork", "Frost Alley", "90210", "United States" for "Jon Snow"
+        And I proceed with "Free" shipping method and "Cash on Delivery" payment
+        And I confirm my order
```
Comment (GSadee) [inline features/checkout/preventing_cart_to_be_modified_after_checkout.feature:22]: As these steps are `Given`:
```suggestion
        And I proceeded with "Free" shipping method and "Cash on Delivery" payment
        And I confirmed my order
```
here and in other scenarios in this file

File: features/checkout/preventing_cart_to_be_modified_after_checkout.feature [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        And the store has a product "AK-47" priced at "$99.99"
+        And the store ships everywhere for free
+        And the store ships everywhere with ups
+        And the store allows paying with "Cash on Delivery"
+        And the store allows paying with "Helicopter Money"
+        And I am a logged in customer
+
+    @api
+    Scenario: Changing address information after completing checkout
+        Given I have product "Sig Sauer P226" in the cart
```
Comment (GSadee) [inline features/checkout/preventing_cart_to_be_modified_after_checkout.feature:19]:
```suggestion
        Given I added product "Sig Sauer P226" to the cart
```
to be consistent with the below steps

File: features/checkout/preventing_cart_to_be_modified_after_checkout.feature [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+        And I specified the billing address as "Ankh Morpork", "Frost Alley", "90210", "United States" for "Jon Snow"
+        And I proceed with "Free" shipping method and "Cash on Delivery" payment
+        And I confirm my order
+        When I try to change the billing address as "Ankh Morpork", "Frost Alley", "90210", "United States" for "Jon Snow"
+        Then I should be informed that cart is no longer available
+
+    @api
+    Scenario: Changing shipping method after completing checkout
```
Comment (GSadee) [inline features/checkout/preventing_cart_to_be_modified_after_checkout.feature:27]:
```suggestion
    Scenario: Preventing from changing shipping method after checkout
```
and a similar change could be applied to other scenarios

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 12]
```diff
@@ -301,6 +317,7 @@ public function iCompleteTheShippingStepWithFirstShippingMethod(): void
      * @When I have proceeded selecting :paymentMethod payment method
      * @When /^the (?:customer|visitor) proceed with ("[^"]+" payment)$/
      * @Given /^the (?:customer|visitor) has proceeded ("[^"]+" payment)$/
+     * @When /^I try to change payment method to ("[^"]+" payment)$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:320]:
```suggestion
     * @When I try to change payment method to :paymentMethod payment
```

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Cart/AddItemToCartHandler.php [hunk 13]
```diff
@@ -57,17 +59,20 @@ public function __construct(
 
     public function __invoke(AddItemToCart $addItemToCart): OrderInterface
     {
+        /** @var ProductVariantInterface $productVariant */
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Cart/AddItemToCartHandler.php:62]:
```suggestion
        /** @var ProductVariantInterface|null $productVariant */
```

**Review by AdamKasp — APPROVED** [review]

File: features/checkout/preventing_cart_from_being_modified_after_checkout.feature [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        And I specified the billing address as "Ankh Morpork", "Frost Alley", "90210", "United States" for "Jon Snow"
+        And I proceeded with "Free" shipping method and "Cash on Delivery" payment
+        And I confirmed my order
+        When I try to add product "AK-47" to the cart
+        Then I should be informed that cart is no longer available
+
+    @api
+    Scenario: Preventing from removing product after checkout
+        Given I added product "Sig Sauer P226" to the cart
+        And I have product "AK-47" in the cart
```
Comment (GSadee) [inline features/checkout/preventing_cart_from_being_modified_after_checkout.feature:56]:
```suggestion
        And I added product "AK-47" to the cart
```

File: features/checkout/preventing_cart_from_being_modified_after_checkout.feature [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+        Then I specified the billing address as "Ankh Morpork", "Frost Alley", "90210", "United States" for "Jon Snow"
+        And I proceeded with "Free" shipping method and "Cash on Delivery" payment
+        And I confirmed my order
+        When I try to remove product "AK-47" from the cart
+        Then I should be informed that cart is no longer available
+
+    @api
+    Scenario: Preventing from changing quantity of product after checkout
+        Given I added product "Sig Sauer P226" to the cart
+        And I have product "AK-47" in the cart
```
Comment (GSadee) [inline features/checkout/preventing_cart_from_being_modified_after_checkout.feature:66]:
```suggestion
        And I added product "AK-47" to the cart
```

**Review by GSadee — APPROVED** [review]

**Comment by AdamKasp** [comment]
Thank you, @arti0090! :tada:

**Pull request merged**